analyzers/src/SonarAnalyzer.CSharp/Rules/MessageTemplates/LoggingTemplatePlaceHoldersShouldBeInOrder.cs

```diff
@@ -0,0 +1,114 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.MessageTemplates;
+
+public sealed class LoggingTemplatePlaceHoldersShouldBeInOrder : IMessageTemplateCheck
+{
+    private const string DiagnosticId = "S6673";
+    private const string MessageFormat = "Template placeholders should be in the right order: placeholder '{0}' does not match with argument '{1}'.";
+
+    internal static readonly DiagnosticDescriptor S6673 = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public DiagnosticDescriptor Rule => S6673;
+
+    public void Execute(SonarSyntaxNodeReportingContext context, InvocationExpressionSyntax invocation, ArgumentSyntax templateArgument, MessageTemplatesParser.Placeholder[] placeholders)
+    {
+        var methodSymbol = (IMethodSymbol)context.SemanticModel.GetSymbolInfo(invocation).Symbol;
+        var templateArguments = PlaceholderValues(invocation, methodSymbol).ToImmutableArray();
```

Comment:
Why `ToImmutableArray` instead of `ToArray` here?

Reply:
It's passed to the `OutOfOrderTemplateArgumentForPlaceHolder` method which takes an `ImmutableArray` parameter (indexing is used on it, so it can't be `IEnumerable`).

Reply:
Still, why not make the argument an array and change the `OutOfOrder` signature?
I don't see a reason to create an immutable array here, since it's all in the same thread.

Reply:
That's a fair point.
I don't mind it, and from looking into it it's not bad performance-wise.
It would be bad if you were appending/prepending to it, but the reason you use it is to signify that you will not mutate it.
Resolving.